src/lib/esnext.iterator.d.ts

```diff
@@ -0,0 +1,125 @@
+/// <reference lib="es2015.iterable" />
+export {};
+
+// Abstract type that allows us to mark `next` as `abstract`
+declare abstract class Iterator<T> {
+    abstract next(value?: undefined): IteratorResult<T, void>;
+}
+
+// Merge all members of `BuiltinIterator<T>` into `Iterator<T>`
+interface Iterator<T> extends globalThis.BuiltinIterator<T, void, undefined> {}
+
+// Capture the `Iterator` constructor in a type we can use in the `extends` clause of `IteratorConstructor`.
+type BuiltinIteratorConstructor = typeof Iterator;
+
+declare global {
+    // Global `BuiltinIterator<T>` interface that can be augmented by polyfills
+    interface BuiltinIterator<T, TReturn, TNext> {
+        /**
+         * Returns this iterator.
+         */
+        [Symbol.iterator](): BuiltinIterator<T, TReturn, TNext>;
+
+        /**
+         * Creates an iterator whose values are the result of applying the callback to the values from this iterator.
+         * @param callbackfn A function that accepts up to two arguments to be used to transform values from the underlying iterator.
+         */
+        map<U>(callbackfn: (value: T, index: number) => U): BuiltinIterator<U>;
+
+        /**
+         * Creates an iterator whose values are those from this iterator for which the provided predicate returns true.
+         * @param predicate A function that accepts up to two arguments to be used to test values from the underlying iterator.
+         */
+        filter<S extends T>(predicate: (value: T, index: number) => value is S): BuiltinIterator<S>;
+
+        /**
+         * Creates an iterator whose values are those from this iterator for which the provided predicate returns true.
+         * @param predicate A function that accepts up to two arguments to be used to test values from the underlying iterator.
+         */
+        filter(predicate: (value: T, index: number) => unknown): BuiltinIterator<T>;
+
+        /**
+         * Creates an iterator whose values are the values from this iterator, stopping once the provided limit is reached.
+         * @param limit The maximum number of values to yield.
+         */
+        take(limit: number): BuiltinIterator<T>;
+
+        /**
+         * Creates an iterator whose values are the values from this iterator after skipping the provided count.
+         * @param count The number of values to drop.
+         */
+        drop(count: number): BuiltinIterator<T>;
+
+        /**
+         * Creates an iterator whose values are the result of applying the callback to the values from this iterator and then flattening the resulting iterators or iterables.
+         * @param callback A function that accepts up to two arguments to be used to transform values from the underlying iterator into new iterators or iterables to be flattened into the result.
+         */
+        flatMap<U>(callback: (value: T, index: number) => Iterator<U> | Iterable<U>): BuiltinIterator<U>;

```

Comment:
Should we restrict this to only allow `Iterator<U, unknown, undefined> | Iterable<U, unknown, undefined>`, since `flatMap` is going to call `next()` on the result without passing an argument?